src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/CertificatePal.cs

```diff
@@ -0,0 +1,538 @@
+                    GC.KeepAlive(this);
+                    return issuerName;
+                }
+            }
+        }
+
+        public IEnumerable<X509Extension> Extensions
+        {
+            get
+            {
+                unsafe
+                {
+                    LowLevelListWithIList<X509Extension> extensions = new LowLevelListWithIList<X509Extension>();
+                    CERT_INFO* pCertInfo = _certContext.CertContext->pCertInfo;
+                    int numExtensions = pCertInfo->cExtension;
+                    for (int i = 0; i < numExtensions; i++)
+                    {
+                        CERT_EXTENSION* pCertExtension = pCertInfo->rgExtension + i;
+                        String oidValue = Marshal.PtrToStringAnsi(pCertExtension->pszObjId);
+                        Oid oid = new Oid(oidValue);
+                        bool critical = pCertExtension->fCritical != 0;
+                        byte[] rawData = pCertExtension->Value.ToByteArray();
+
+                        X509Extension extension = new X509Extension(oid, rawData, critical);
+                        extensions.Add(extension);
+                    }
+                    GC.KeepAlive(this);
+                    return extensions;
+                }
+            }
+        }
+
+        public String GetNameInfo(X509NameType nameType, bool forIssuer)
+        {
+            CertNameType certNameType = MapNameType(nameType);
+            CertNameFlags certNameFlags = forIssuer ? CertNameFlags.CERT_NAME_ISSUER_FLAG : CertNameFlags.None;
+            CertNameStrTypeAndFlags strType = CertNameStrTypeAndFlags.CERT_X500_NAME_STR | CertNameStrTypeAndFlags.CERT_NAME_STR_REVERSE_FLAG;
+
+            int cchCount = Interop.crypt32.CertGetNameString(_certContext, certNameType, certNameFlags, ref strType, null, 0);
+            if (cchCount == 0)
+                throw new CryptographicException(Marshal.GetLastWin32Error());
+
+            StringBuilder sb = new StringBuilder(cchCount);
+            if (Interop.crypt32.CertGetNameString(_certContext, certNameType, certNameFlags, ref strType, sb, cchCount) == 0)
+                throw new CryptographicException(Marshal.GetLastWin32Error());
+
+            return sb.ToString();
+        }
+
+        public void AppendPrivateKeyInfo(StringBuilder sb)
+        {
+            CspKeyContainerInfo cspKeyContainerInfo = null;
+            try
+            {
+                if (this.HasPrivateKey)
+                {
+                    CspParameters parameters = GetPrivateKey();
+                    cspKeyContainerInfo = new CspKeyContainerInfo(parameters);
+                }
+            }
+            // We could not access the key container. Just return.
+            catch (CryptographicException) { }
+
+            if (cspKeyContainerInfo == null)
+                return;
+
+            sb.Append(Environment.NewLine + Environment.NewLine + "[Private Key]");
```

Comment:
So many unnecessary interim string allocations :(  No idea how much this function matters perf-wise, but at some point it'd be good to break these apart to avoid the extra allocations.